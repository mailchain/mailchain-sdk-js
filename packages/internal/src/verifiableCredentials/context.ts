export const W3C_CREDENTIALS_CONTEXT = 'https://www.w3.org/2018/credentials/v1';
