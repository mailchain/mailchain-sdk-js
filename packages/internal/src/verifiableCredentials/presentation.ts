import { PresentationPayload, VerifiableCredential } from 'did-jwt-vc';
import { DecentralizedIdentifier, MailchainDecentralizedIdentifier } from './did';
import { W3C_CREDENTIALS_CONTEXT } from './context';

type CreatePresentationPayloadParams = {
	id?: string;
	verifiableCredential: VerifiableCredential;
	verifier: DecentralizedIdentifier;
	issuanceDate: Date;
	/**
	 * The date and time that the credential will expire.
	 */
	expirationDate?: Date;
	holder: MailchainDecentralizedIdentifier;
};

/**
 * Create presentation payload that is then signed to create a verified presentation.
 * @param params
 * @returns
 */
export function createPresentationPayload(params: CreatePresentationPayloadParams): PresentationPayload {
	const { verifiableCredential, verifier, issuanceDate, expirationDate, holder, id } = params;

	return {
		'@context': [W3C_CREDENTIALS_CONTEXT],
		type: ['VerifiablePresentation'],
		id,
		holder,
		verifiableCredential: [verifiableCredential],
		verifier,
		issuanceDate: issuanceDate.toUTCString(),
		expirationDate: expirationDate ? expirationDate.toUTCString() : undefined,
	};
}
