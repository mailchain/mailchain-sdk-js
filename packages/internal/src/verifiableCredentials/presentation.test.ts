import { W3C_CREDENTIALS_CONTEXT } from './context';
import { createPresentationPayload } from './presentation';

describe('createPresentationPayload', () => {
	it('create', () => {
		const actual = createPresentationPayload({
			holder: 'did:mailchain:holder',
			issuanceDate: new Date(2020, 1, 1),
			verifiableCredential: {
				'@context': [W3C_CREDENTIALS_CONTEXT],
				credentialSubject: {
					id: 'credentialSubjectId',
				},
				issuanceDate: new Date(2000, 1, 1).toISOString(),
				issuer: { id: 'did:mailchain:issuer' },
				proof: { id: 'proof' },
				type: ['type'],
			},
			verifier: 'did:mailchain:verifier',
		});

		expect(actual).toMatchSnapshot();
	});
});
