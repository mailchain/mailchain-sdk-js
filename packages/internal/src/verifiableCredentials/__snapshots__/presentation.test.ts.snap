// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPresentationPayload create 1`] = `
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "expirationDate": undefined,
  "holder": "did:mailchain:holder",
  "issuanceDate": "Sat, 01 Feb 2020 00:00:00 GMT",
  "requestId": "request-id",
  "type": [
    "VerifiablePresentation",
  ],
  "verifiableCredential": [
    {
      "@context": [
        "https://www.w3.org/2018/credentials/v1",
      ],
      "credentialSubject": {
        "id": "credentialSubjectId",
      },
      "issuanceDate": "2000-02-01T00:00:00.000Z",
      "issuer": {
        "id": "did:mailchain:issuer",
      },
      "proof": {
        "id": "proof",
      },
      "type": [
        "type",
      ],
    },
  ],
  "verifier": "did:mailchain:verifier",
}
`;
